### PR TITLE
#484 Make scrolling through the whole file content possible with a frontend cache in the UI fabric List element

### PR DESCRIPTION
**Pull request by MelindaSW**
### In this PR

image

The custom slider scrollbar is removed and the frontend, and some parts of the backend, is adapted to use the "normal" scroller and still be able to scroll through the whole file. 

When opening a file, a part of the beginning of the file is added to the backend cache. 
Half of it is sent to the frontend. Approximately 500 lines when reading from a spring log.

In order to make the scrollbar scale, the total line amount of the file is then calculated. 
Once the line count arrives in the frontend, it is used to expand the list with undefined values in the array that gets passed into the UI Fabric List component. The List renders them as visually empty loglines. 
We tried it with other options, like UI Fabric shimmer element, but decided to go with this solution for now.

Scrolling past a certain amount of items triggers a request of new lines from the backend. 
If the total file content is already in the frontend, a request is never triggered.

The engine gets the request and first checks the backend cache for the content. It it is there it returns an array of new lines. 
If the lines are not found it updates the cache with new content from the file and makes another search.

The LogViewerReducer replaces the old log lines with the new.
They go through the process in the mainScriptOffloader before they are inserted into the array with undefined values again.

### Try it out

Curious now, and a bit nervous, to know how it works on your machines.

If you are opening a bigger file, wait until the line count is finished before you start scrolling. At the moment we have not implemented any indicator or spinner signalling that we should wait. This could perhaps be a separate issue as we are unsure of what would be the best looking solution. 
Instead, keep an eye out for ADDING LINE COUNT to appear in the console, then start scrolling.  

image

We have tried it with files of different sizes and it seems to work well with fairly large files, up to 1GB. The behaviour of the list is a little weird when working with a 2.5gb logfile, especially when scrolling to the end, although nothing is crashing. Let us know if it happens for you as well. For example, look at the wonky line in the middle here.

image

### Features still needing some adjustments to work with the new implementation

-	Filtering, #460
-	Following a file, #505 

### Issues related to the previous solution. Can be removed if this PR is approved

#480, #487, #467, #477

### CODE REVIEW

File: src/js/engine/engine.js [hunk 1]
```diff
@@ -26,19 +24,15 @@ const getFileInfo = async filePath => {
...
     lines,
     linesStartAt,
     linesEndAt
-  } = await fileReader.readDataFromByte(
-    filePath,
-    START_READ_FROM_BYTE,
-    NR_OF_BYTES
-  );
+  } = await fileReader.readDataFromByte(filePath, startFromByte, nrOfBytes);
+
```
Comment (Atmanea) [inline src/js/engine/engine.js:35]: Bara nyfiken här om det är någon särskild anledning varför variabelnamnen gick från stora bokstäver till små i och med att de är konstanter?

Comment (MelindaSW) [reply src/js/engine/engine.js:35]: Hmm, det var nog, om jag minns rätt, för att jag var osäker på konventioner i javascript när jag skrev dem med stora bokstäver en gång i tiden. Bestämde mig för att ändra till camel case för att det matchade hur andra har gjort i koden i övrigt.

Comment (Atmanea) [reply src/js/engine/engine.js:35]: Ok, makes sense, tack!

File: src/js/engine/engine.js [hunk 2]
```diff
@@ -68,47 +61,46 @@ const sendFileOpened = async (
...
+        type: 'TOTAL_LINE_AMOUNT_CALCULATED',
+        data: {
+          filePath,
+          lineCount
+        }
+      };
+      sender.send(ipcChannel, action);
+    })
+    .catch(err => {
+      console.log({ sendTotalLineCount, err });
```
Comment (Atmanea) [inline src/js/engine/engine.js:86]: Maybe it is a better idea to use await here for readability?

File: src/js/hiddenWindow/mainScriptOffloader.js [hunk 3]
```diff
@@ -101,6 +90,37 @@ class LogsFiltererAndHighlighter {
   }
 }
 
+const filterExistsAndMatchesWithLineOrHasFilterNotBeenSet = (
```
Comment (Atmanea) [inline src/js/hiddenWindow/mainScriptOffloader.js:93]: Gillar den nya ändringen med att ta ut checkarna till egna funktioner och sätta beskrivande namn på dom. Det gör det enklare att förstå vad som händer och mindre 'kluddigare' i start-funktionen.

**Pull request merged**